### PR TITLE
mixedversion: introduce cluster setting mutator

### LINKED ISSUE #114723: mixedversion: support setting certain cluster settings randomly
It would be a good feature to support randomly setting certain cluster settings that are expected to have no impact on _most_ tests. Settings would be introduced at a random point in the test (whenever they are available) and could optionally be unset as well.

One example of such setting is `kv.expiration_leases_only.enabled` (see #112226). Tests would have the option to disable this behaviour of the test if what it does is incompatible with the cluster setting.

### Other options considered:

* **Making this a roachtest default behaviour**: this would be ideal (as it would apply to every test, not just mixedversion tests). In fact, there's already an API to do this for the expiration-based leases mentioned above (see `TestSpec.ExpirationLeases`). However, roachtest does not currently have the notion of "version" when starting cockroach processes. As a consequence, we can't reliably use that API since the version we are starting might not support the cluster setting.

* **Relying on metamorphic builds of cockroach**: This should eventually be supported, but there's still work to be done to fully integrate metamorphic builds in regular roachtest runs. Supporting them in mixedversion is still in the more distant future, so it will be nice to have something that works until we get there.

Jira issue: CRDB-33654

### PR DESCRIPTION
**Pull request by renatolabs**
This commit introduces both a generic set of functions that can be
used to implement mutators for various cluster settings, and
introduces a new cluster setting mutator for the
`kv.expiration_leases_only.enabled` cluster setting.

The mutator implementation itself is agnostic of the cluster setting
it is changing. It takes the cluster setting name, possible values it
can assume, number of changes (etc), and generates mutations that will
non-deterministicall set (and reset) the cluster setting during the
mixedversion test.

Fixes: #114723

Release note: None

### CODE REVIEW

**Comment by renatolabs** [comment]
> Your pull request contains more than 1000 changes. It is strongly encouraged to split big PRs into smaller chunks.

In my defense, the first commit is purely mechanical and accounts for half of the diff (hence, reviewing by commit is highly encouraged).

I ran all mixed-version tests on this branch 3 times forcing the new mutator to be enabled [and everything went well](https://teamcity.cockroachdb.com/viewLog.html?buildId=14450968&buildTypeId=Cockroach_Nightlies_RoachtestNightlyGceBazel&tab=buildResultsDiv&branch_Cockroach_Nightlies=120559).

This is ready for a serious look. 🙏

File: pkg/cmd/roachtest/roachtestutil/mixedversion/mutators_test.go [hunk 1]
```diff
@@ -45,3 +51,200 @@ func TestPreserveDowngradeOptionRandomizerMutator(t *testing.T) {
...
 	}
 }
+
+// TestClusterSettingMutator does not validate the specific mutations
+// generated by the clusterSettingMutartor; instead, it validates the
+// invariants that the mutator should provide. For example: expected
+// number of mutations, no consecutive RESETs, no consecutive SETs to
+// same value, etc. This is done for different configurations of the
+// mutator.
+func TestClusterSettingMutator(t *testing.T) {
```
Comment (herkolategan) [inline pkg/cmd/roachtest/roachtestutil/mixedversion/mutators_test.go:61]: Nice!

**Review by herkolategan — APPROVED** [review]
Nice work!

**Comment by renatolabs** [comment]
Ran a few mixed-version tests on this branch to make sure things continue to work well: [link](https://teamcity.cockroachdb.com/viewLog.html?buildId=14571591&buildTypeId=Cockroach_Nightlies_RoachtestNightlyGceBazel&tab=buildResultsDiv).

TFTR!

bors r=herkolategan

**Comment by renatolabs** [comment]
This won't work, I'll try anyway:

blathers backport 23.2 23.1

**Pull request merged**